File: apps/demo-be-python-fastapi/tests/integration/steps/admin_steps.py
"""BDD step definitions for admin feature."""

from pytest_bdd import given, parsers, scenarios, then, when

from tests.integration.conftest import GHERKIN_ROOT
from tests.integration.service_client import FakeResponse, ServiceClient
from tests.integration.steps.security_steps import _ADMIN_PASSWORD, _register_and_promote_admin

scenarios(str(GHERKIN_ROOT / "admin" / "admin.feature"))

_PASSWORD = "Str0ng#Pass1"


@given(
    parsers.parse('an admin user "{username}" is registered and logged in'),
    target_fixture="admin_tokens",
)
def admin_login(client: ServiceClient, username: str) -> dict:
    user_data = _register_and_promote_admin(client, username, _ADMIN_PASSWORD)
    tokens = client.login_user(username, _ADMIN_PASSWORD)
    return {**tokens, "admin_id": user_data["id"]}


@given(
    parsers.parse('users "{a}", "{b}", and "{c}" are registered'),
    target_fixture="registered_users",
)
def register_multiple_users(client: ServiceClient, a: str, b: str, c: str) -> list:
    users = []
    for username in [a, b, c]:
        users.append(client.register_user(username))
    return users


@given(
    parsers.parse('"{username}" has logged in and stored the access token'),
    target_fixture="alice_tokens",
)
def alice_login_for_admin(client: ServiceClient, username: str) -> dict:
    return client.login_user(username, _PASSWORD)


@given("alice's account has been disabled by the admin")
def disable_alice_by_admin(
    client: ServiceClient, registered_users: list, admin_tokens: dict
) -> None:
    alice = next((u for u in registered_users if u["username"] == "alice"), None)
    assert alice is not None
    resp = client.post_admin_disable_user(
        alice["id"],
        f"Bearer {admin_tokens['access_token']}",
        reason="Test disable",
    )
    assert resp.status_code == 200


@given("alice's account has been disabled")
def alice_account_disabled(
    client: ServiceClient, registered_users: list, admin_tokens: dict
) -> None:
    alice = next((u for u in registered_users if u["username"] == "alice"), None)
    assert alice is not None
    resp = client.post_admin_disable_user(
        alice["id"],
        f"Bearer {admin_tokens['access_token']}",
        reason="Initial disable",
    )
    assert resp.status_code == 200


# --- When steps ---


@when("the admin sends GET /api/v1/admin/users", target_fixture="response")
def admin_list_users(client: ServiceClient, admin_tokens: dict) -> FakeResponse:
    return client.get_admin_users(f"Bearer {admin_tokens['access_token']}")


@when(
    "the admin sends GET /api/v1/admin/users?email=alice@example.com",
    target_fixture="response",
)
def admin_search_users_by_email(client: ServiceClient, admin_tokens: dict) -> FakeResponse:
    return client.get_admin_users(
        f"Bearer {admin_tokens['access_token']}",
        email="alice@example.com",
    )


@when(
    parsers.parse("the admin sends POST /api/v1/admin/users/{{alice_id}}/disable with body {body}"),
    target_fixture="response",
)
def admin_disable_alice(
    client: ServiceClient, registered_users: list, admin_tokens: dict, body: str
) -> FakeResponse:
    import json

    alice = next((u for u in registered_users if u["username"] == "alice"), None)
    assert alice is not None
    data = json.loads(body)
    return client.post_admin_disable_user(
        alice["id"],
        f"Bearer {admin_tokens['access_token']}",
        reason=data.get("reason"),
    )


@when(
    "the client sends GET /api/v1/users/me with alice's access token",
    target_fixture="response",
)
def get_me_alice_token(client: ServiceClient, alice_tokens: dict) -> FakeResponse:
    return client.get_me(f"Bearer {alice_tokens['access_token']}")


@when(
    parsers.parse("the admin sends POST /api/v1/admin/users/{{alice_id}}/enable"),
    target_fixture="response",
)
def admin_enable_alice(
    client: ServiceClient, registered_users: list, admin_tokens: dict
) -> FakeResponse:
    alice = next((u for u in registered_users if u["username"] == "alice"), None)
    assert alice is not None
    return client.post_admin_enable_user(
        alice["id"],
        f"Bearer {admin_tokens['access_token']}",
    )


@when(
    parsers.parse("the admin sends POST /api/v1/admin/users/{{alice_id}}/force-password-reset"),
    target_fixture="response",
)
def admin_force_password_reset(
    client: ServiceClient, registered_users: list, admin_tokens: dict
) -> FakeResponse:
    alice = next((u for u in registered_users if u["username"] == "alice"), None)
    assert alice is not None
    return client.post_admin_force_password_reset(
        alice["id"],
        f"Bearer {admin_tokens['access_token']}",
    )


# --- Then steps ---


@then(
    'the response body should contain at least one user with "email" equal to "alice@example.com"'
)
def check_alice_in_results(response: FakeResponse) -> None:
    body = response.json()
    users = body.get("data", [])
    assert any(u.get("email") == "alice@example.com" for u in users), (
        f"alice@example.com not found in users: {users}"
    )


@then('alice\'s account status should be "disabled"')
def check_alice_disabled(
    client: ServiceClient, registered_users: list, admin_tokens: dict
) -> None:
    alice = next((u for u in registered_users if u["username"] == "alice"), None)
    assert alice is not None
    resp = client.get_admin_users(
        f"Bearer {admin_tokens['access_token']}",
        email="alice@example.com",
    )
    body = resp.json()
    users = body.get("data", [])
    alice_info = next((u for u in users if u["username"] == "alice"), None)
    assert alice_info is not None
    assert alice_info["status"].upper() == "DISABLED"


@then('alice\'s account status should be "active"')
def check_alice_active(
    client: ServiceClient, registered_users: list, admin_tokens: dict
) -> None:
    alice = next((u for u in registered_users if u["username"] == "alice"), None)
    assert alice is not None
    resp = client.get_admin_users(
        f"Bearer {admin_tokens['access_token']}",
        email="alice@example.com",
    )
    body = resp.json()
    users = body.get("data", [])
    alice_info = next((u for u in users if u["username"] == "alice"), None)
    assert alice_info is not None
    assert alice_info["status"].upper() == "ACTIVE"

File: apps/demo-be-rust-axum/tests/unit/steps/admin_steps.rs
use cucumber::{given, then, when};

use crate::world::{get_req, json_req, AppWorld};

#[given(expr = "users {string}, {string}, and {string} are registered")]
async fn register_multiple_users(world: &mut AppWorld, u1: String, u2: String, u3: String) {
    for (username, email) in [
        (u1.as_str(), format!("{u1}@example.com")),
        (u2.as_str(), format!("{u2}@example.com")),
        (u3.as_str(), format!("{u3}@example.com")),
    ] {
        let body = format!(
            r#"{{"username": "{username}", "email": "{email}", "password": "Str0ng#Pass1"}}"#
        );
        let req = json_req("POST", "/api/v1/auth/register", &body, None);
        world.send(req).await.unwrap();
        if world.last_status == 201 && username == u1.as_str() {
            world.alice_id = world
                .last_body
                .get("id")
                .and_then(|v| v.as_str())
                .and_then(|s| uuid::Uuid::parse_str(s).ok());
        }
    }
}

#[when("the admin sends GET /api/v1/admin/users")]
async fn admin_list_users(world: &mut AppWorld) {
    let bearer = world.admin_bearer();
    let req = get_req("/api/v1/admin/users", Some(&bearer));
    world.send(req).await.unwrap();
}

#[when(regex = r#"the admin sends GET /api/v1/admin/users\?search=alice@example\.com"#)]
async fn admin_search_users_by_email(world: &mut AppWorld) {
    let bearer = world.admin_bearer();
    let req = get_req(
        "/api/v1/admin/users?search=alice@example.com",
        Some(&bearer),
    );
    world.send(req).await.unwrap();
}

#[then(expr = "the response body should contain at least one user with {string} equal to {string}")]
async fn response_has_user_with_field(world: &mut AppWorld, field: String, value: String) {
    let data = world.last_body.get("content").and_then(|v| v.as_array());
    let found = data
        .map(|arr| {
            arr.iter().any(|user| {
                user.get(&field)
                    .and_then(|v| v.as_str())
                    .map(|s| s == value.as_str())
                    .unwrap_or(false)
            })
        })
        .unwrap_or(false);
    assert!(
        found,
        "Expected at least one user with '{field}' = '{value}' in: {}",
        world.last_body
    );
}

#[when(
    regex = r#"the admin sends POST /api/v1/admin/users/\{alice_id\}/disable with body \{ "reason": "Policy violation" \}"#
)]
async fn admin_disable_alice(world: &mut AppWorld) {
    let bearer = world.admin_bearer();
    let alice_id = world.alice_id.map(|id| id.to_string()).unwrap_or_default();
    let req = json_req(
        "POST",
        &format!("/api/v1/admin/users/{alice_id}/disable"),
        r#"{"reason": "Policy violation"}"#,
        Some(&bearer),
    );
    world.send(req).await.unwrap();
}

#[given("alice's account has been disabled by the admin")]
async fn alice_account_disabled_by_admin(world: &mut AppWorld) {
    let bearer = world.admin_bearer();
    let alice_id = world.alice_id.map(|id| id.to_string()).unwrap_or_default();
    if !alice_id.is_empty() {
        let req = json_req(
            "POST",
            &format!("/api/v1/admin/users/{alice_id}/disable"),
            r#"{"reason": "Policy violation"}"#,
            Some(&bearer),
        );
        world.send(req).await.unwrap();
    }
}

#[given("alice's account has been disabled")]
async fn alice_account_disabled_simple(world: &mut AppWorld) {
    let bearer = world.admin_bearer();
    let alice_id = world.alice_id.map(|id| id.to_string()).unwrap_or_default();
    if !alice_id.is_empty() && !bearer.is_empty() {
        let req = json_req(
            "POST",
            &format!("/api/v1/admin/users/{alice_id}/disable"),
            r#"{"reason": "test"}"#,
            Some(&bearer),
        );
        world.send(req).await.unwrap();
    }
}

#[when("the admin sends POST /api/v1/admin/users/{alice_id}/enable")]
async fn admin_enable_alice(world: &mut AppWorld) {
    let bearer = world.admin_bearer();
    let alice_id = world.alice_id.map(|id| id.to_string()).unwrap_or_default();
    let req = json_req(
        "POST",
        &format!("/api/v1/admin/users/{alice_id}/enable"),
        "{}",
        Some(&bearer),
    );
    world.send(req).await.unwrap();
}

#[then(expr = "alice's account status should be {string}")]
async fn alice_account_status(world: &mut AppWorld, expected: String) {
    let expected_lower = expected.to_lowercase();
    match expected_lower.as_str() {
        "active" | "disabled" => {
            assert_eq!(
                world.last_status, 200,
                "Expected 200 for status action, got {}",
                world.last_status
            );
        }
        "locked" => {
            assert_eq!(
                world.last_status, 401,
                "Expected 401 for locked account login, got {}",
                world.last_status
            );
        }
        _ => {}
    }
}

#[when("the admin sends POST /api/v1/admin/users/{alice_id}/force-password-reset")]
async fn admin_force_password_reset(world: &mut AppWorld) {
    let bearer = world.admin_bearer();
    let alice_id = world.alice_id.map(|id| id.to_string()).unwrap_or_default();
    let req = json_req(
        "POST",
        &format!("/api/v1/admin/users/{alice_id}/force-password-reset"),
        "{}",
        Some(&bearer),
    );
    world.send(req).await.unwrap();
}
